authz/grpc_audit_end2end_test.go

```diff
@@ -0,0 +1,332 @@
+/*
+ *
+ * Copyright 2023 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package authz_test
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"io"
+	"net"
+	"os"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/authz"
+	"google.golang.org/grpc/authz/audit"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/internal/grpctest"
+	testgrpc "google.golang.org/grpc/interop/grpc_testing"
+	testpb "google.golang.org/grpc/interop/grpc_testing"
+	"google.golang.org/grpc/testdata"
+
+	_ "google.golang.org/grpc/authz/audit/stdout"
+)
+
+func TestAudit(t *testing.T) {
+	grpctest.RunSubTests(t, s{})
+}
+
+type statAuditLogger struct {
+	AuthzDescisionStat map[bool]int      // Map to hold the counts of authorization decisions
+	EventContent       map[string]string // Map to hold event fields in key:value fashion
+	SpiffeIds          []string          // Slice to hold collected SPIFFE IDs
+}
+
+func (s *statAuditLogger) Log(event *audit.Event) {
+	s.AuthzDescisionStat[event.Authorized]++
+	s.EventContent["rpc_method"] = event.FullMethodName
+	s.EventContent["principal"] = event.Principal
+	s.EventContent["policy_name"] = event.PolicyName
+	s.EventContent["matched_rule"] = event.MatchedRule
+	s.EventContent["authorized"] = strconv.FormatBool(event.Authorized)
+}
+
+type loggerBuilder struct {
+	AuthDecisionStat map[bool]int
+	EventContent     map[string]string
+}
+
+func (loggerBuilder) Name() string {
+	return "stat_logger"
+}
+func (lb *loggerBuilder) Build(audit.LoggerConfig) audit.Logger {
+	return &statAuditLogger{
+		AuthzDescisionStat: lb.AuthDecisionStat,
+		EventContent:       lb.EventContent,
+	}
+}
+
+func (*loggerBuilder) ParseLoggerConfig(config json.RawMessage) (audit.LoggerConfig, error) {
+	return nil, nil
+}
+
+func (s) TestAuditLogger(t *testing.T) {
+	tests := map[string]struct {
```

Comment:
Nit: I'm not sure I like this pattern all that much (`map` instead of a slice), and I don't think I've seen it before, either.  It forces even more indentation than there already is, and that makes the test cases harder to read.  Typically we do:

```go
tests := []struct{
  name string
  .....
}{
    {
        name: "..."
    },
    {
        name: "..."
    }
}

// OR tighter spacing which I prefer but others may not.

tests := []struct{
  name string
  .....
}{{
    name: "..."
}, {
    name: "..."
}}
```